File: sdk-python/src/personaai/resources/skills.py
"""Skills (``/api/v1/developer/skills``) — Project-owned or, when this
client asserts an external user, owned by that end user. Ported from
``sdk/src/resources/skills.ts``."""

from __future__ import annotations

from typing import TYPE_CHECKING, cast

from ..types.skill import CreateSkillInput, DiscoverSkillsParams, Skill, UpdateSkillInput
from ..types.usage import ResourceUsage

if TYPE_CHECKING:
    from .._async_http import AsyncTransport
    from .._sync_http import SyncTransport


class Skills:
    def __init__(self, transport: SyncTransport) -> None:
        self._transport = transport

    def create(self, input: CreateSkillInput) -> Skill:
        return cast(Skill, self._transport.request("POST", "/api/v1/developer/skills", json=input))

    def list(self, params: DiscoverSkillsParams | None = None) -> list[Skill]:
        """Note: returns a bare list — this endpoint has no pagination envelope."""
        return cast(
            list[Skill],
            self._transport.request("GET", "/api/v1/developer/skills", query=params),
        )

    def get(self, skill_id: str) -> Skill:
        return cast(Skill, self._transport.request("GET", f"/api/v1/developer/skills/{skill_id}"))

    def update(self, skill_id: str, input: UpdateSkillInput) -> Skill:
        return cast(
            Skill,
            self._transport.request("PATCH", f"/api/v1/developer/skills/{skill_id}", json=input),
        )

    def delete(self, skill_id: str) -> None:
        self._transport.request("DELETE", f"/api/v1/developer/skills/{skill_id}")

    def get_usage(self, skill_id: str) -> ResourceUsage:
        """Agents referencing this Skill — check before ``delete()`` to avoid a blocked delete."""
        return cast(
            ResourceUsage,
            self._transport.request("GET", f"/api/v1/developer/skills/{skill_id}/usage"),
        )


class AsyncSkills:
    def __init__(self, transport: AsyncTransport) -> None:
        self._transport = transport

    async def create(self, input: CreateSkillInput) -> Skill:
        return cast(
            Skill, await self._transport.request("POST", "/api/v1/developer/skills", json=input)
        )

    async def list(self, params: DiscoverSkillsParams | None = None) -> list[Skill]:
        """Note: returns a bare list — this endpoint has no pagination envelope."""
        return cast(
            list[Skill],
            await self._transport.request("GET", "/api/v1/developer/skills", query=params),
        )

    async def get(self, skill_id: str) -> Skill:
        return cast(
            Skill, await self._transport.request("GET", f"/api/v1/developer/skills/{skill_id}")
        )

    async def update(self, skill_id: str, input: UpdateSkillInput) -> Skill:
        return cast(
            Skill,
            await self._transport.request(
                "PATCH", f"/api/v1/developer/skills/{skill_id}", json=input
            ),
        )

    async def delete(self, skill_id: str) -> None:
        await self._transport.request("DELETE", f"/api/v1/developer/skills/{skill_id}")

    async def get_usage(self, skill_id: str) -> ResourceUsage:
        """Agents referencing this Skill — check before ``delete()`` to avoid a blocked delete."""
        return cast(
            ResourceUsage,
            await self._transport.request("GET", f"/api/v1/developer/skills/{skill_id}/usage"),
        )

File: sdk/src/resources/skills.ts
import type { HttpClient } from '../http.js';
import type {
  CreateSkillInput,
  DiscoverSkillsParams,
  Skill,
  UpdateSkillInput,
} from '../types/skill.js';
import type { ResourceUsage } from '../types/usage.js';
import type { BulkDeleteResult } from '../types/bulkDelete.js';
import type { PaginatedResult } from '../types/pagination.js';

/**
 * Skills (`/api/v1/developer/skills`) — Project-owned or, when this client
 * asserts an external user, owned by that end user.
 */
export class SkillsResource {
  constructor(private readonly http: HttpClient) {}

  /**
   * `idempotencyKey`, if provided, is sent as the `Idempotency-Key` header —
   * a safe retry with the same key replays the original response instead
   * of creating a duplicate Skill.
   */
  async create(input: CreateSkillInput, idempotencyKey?: string): Promise<Skill> {
    return this.http.request<Skill>('POST', '/api/v1/developer/skills', {
      body: input,
      headers: idempotencyKey ? { 'Idempotency-Key': idempotencyKey } : undefined,
    });
  }

  async list(params: DiscoverSkillsParams = {}): Promise<PaginatedResult<Skill>> {
    return this.http.request<PaginatedResult<Skill>>('GET', '/api/v1/developer/skills', {
      query: { ...params },
    });
  }

  async get(skillId: string): Promise<Skill> {
    return this.http.request<Skill>('GET', `/api/v1/developer/skills/${skillId}`);
  }

  async update(skillId: string, input: UpdateSkillInput): Promise<Skill> {
    return this.http.request<Skill>('PATCH', `/api/v1/developer/skills/${skillId}`, {
      body: input,
    });
  }

  async delete(skillId: string): Promise<void> {
    await this.http.request<unknown>('DELETE', `/api/v1/developer/skills/${skillId}`);
  }

  /** Agents referencing this Skill — check before `delete()` to avoid a blocked-delete error. */
  async getUsage(skillId: string): Promise<ResourceUsage> {
    return this.http.request<ResourceUsage>('GET', `/api/v1/developer/skills/${skillId}/usage`);
  }

  /** Best-effort batch delete — up to 100 ids per call; partial failures don't throw. */
  async bulkDelete(ids: string[]): Promise<BulkDeleteResult> {
    return this.http.request<BulkDeleteResult>('POST', '/api/v1/developer/skills/bulk-delete', {
      body: { ids },
    });
  }
}
